Use FileStore in TorchScript for store registry

wanchaol opened the pull request:
Summary: make the FileStore path also use TorchScript when it's needed.

Test Plan: wait for sandcastle.

Reviewed By: zzzwen

Differential Revision: D25842651

pritamdamania87 commented on torch/csrc/distributed/c10d/init.cpp:
```diff
@@ -1255,6 +1255,14 @@ that adds a prefix to each key inserted to the store.
 static const auto StoreTorchBind =
     torch::class_<::c10d::Store>("dist_c10d", "Store");
 
+static const auto FileStoreTorchBind =
+    torch::class_<::c10d::FileStore>("dist_c10d", "FileStore")
+        .def(torch::init([](const std::string& path,
+                            int64_t num_workers) {
+          return c10::make_intrusive<::c10d::FileStore>(
+              path, num_workers);
+        }));
+
```
Can we add some unit tests verifying this works as expected in TorchScript? Also, I just noticed that users would have to use something like `torch.classes.dist_c10d.Filestore` to use this in TorchScript. Is it possible to bind `torch.distributed.FileStore` to this class instead when we are under TorchScript?

wanchaol replied:
Yes, the future plan is to use TorchBind (instead of pybind11) to bind the custom class like `FileStore` directly, but currently there're bunch of improvements need to be done between we fully switch to it. Right now this binding co-exist with pybind bindings, and unfortunately user have to use this initially as they are not convertible between each other. We will do a full switch to this binding once we finish the improvements to TorchBind, after that user can automatically use the same API like `torch.distributed.FileStore` inside TorchScript.

pritamdamania87 commented on test/distributed/test_jit_c10d.py:
```diff
@@ -144,6 +144,20 @@ def forward(self, input: torch.Tensor):
+    def setUp(self):
+        super(StoreTest, self).setUp()
+        self.file = tempfile.NamedTemporaryFile(delete=False)
+        self.filestore = torch.classes.dist_c10d.FileStore(self.file.name, 1)
```
Shouldn't we test that calling this within JIT works appropriately? Or is there some magic in `JitTestCase` that is already doing this?

wanchaol replied:
nothing magic, the internal use case right now only testing if python side works when creating the process group, but it's a better idea that we also test the JIT side store creation works fine, updating.

Commit:
Use FileStore in TorchScript for store registry (#50248)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/50248

make the FileStore path also use TorchScript when it's needed.

Test Plan: wait for sandcastle.

Reviewed By: zzzwen

Differential Revision: D25842651

fbshipit-source-id: 08d277f32f01f2ac4e105fdbf74f76b839daaa29

gmagogsfm commented on test/distributed/test_jit_c10d.py:
```diff
@@ -144,6 +144,36 @@ def forward(self, input: torch.Tensor):
+        def create_file_store(
+            path: str,
+            num_workers: int
+        ) -> torch.classes.dist_c10d.FileStore:
```
Are there cases where we need this in eager? if so, maybe cover that as well?

wanchaol replied:
Yes actually the creation logic in eager already been covered in `setUp` and been tested in the `test_create_prefix_store`
